pandablocks/ioc/ioc.py

```diff
@@ -0,0 +1,1756 @@
+
+        max_delay_record_name = EpicsName(record_name + ":MAX_DELAY")
+        record_dict[max_delay_record_name] = self._create_record_info(
+            max_delay_record_name,
+            "Maximum valid input delay",
+            builder.longIn,
+            type(field_info.max_delay),
+            initial_value=field_info.max_delay,
+        )
+
+        return record_dict
+
+    def _make_pos_mux(
+        self,
+        record_name: EpicsName,
+        field_info: FieldInfo,
+        values: Dict[EpicsName, ScalarRecordValue],
+    ) -> Dict[EpicsName, RecordInfo]:
+
+        self._check_num_values(values, 1)
+        assert isinstance(field_info, PosMuxFieldInfo)
+        assert field_info.labels
+
+        record_dict: Dict[EpicsName, RecordInfo] = {}
+
+        # This should be an mbbOut record, but there are too many posssible labels
+        # TODO: There will need to be some mechanism to retrieve the labels,
+        # but there's a POSITIONS table that can probably be used
+        validator = StringRecordLabelValidator(field_info.labels)
+        # Ensure we're putting a valid value to start with
+        assert values[record_name] in field_info.labels
+
+        record_dict[record_name] = self._create_record_info(
+            record_name,
+            field_info.description,
+            builder.stringOut,
+            str,
+            initial_value=values[record_name],
+            validate=validator.validate,
+        )
+
+        return record_dict
+
+    def _make_table(
+        self,
+        record_name: EpicsName,
+        field_info: FieldInfo,
+        values: Dict[EpicsName, List[str]],
+    ) -> Dict[EpicsName, RecordInfo]:
+        assert isinstance(field_info, TableFieldInfo)
+
+        table_updater = TableUpdater(
+            self._client,
+            record_name,
+            field_info,
+            self._all_values_dict,
+        )
+
+        # Format the mode record name to remove namespace
+        mode_record_name: str = table_updater.mode_record_info.record.name
+        mode_record_name = EpicsName(
+            mode_record_name.replace(self._record_prefix + ":", "")
+        )
+
+        return {mode_record_name: table_updater.mode_record_info}
+
+    def _make_uint(
+        self,
+        record_name: EpicsName,
+        field_info: FieldInfo,
+        record_creation_func: Callable,
+        **kwargs,
+    ) -> Dict[EpicsName, RecordInfo]:
+        assert isinstance(field_info, UintFieldInfo)
+
+        record_dict: Dict[EpicsName, RecordInfo] = {}
+        record_dict[record_name] = self._create_record_info(
+            record_name,
+            field_info.description,
+            record_creation_func,
+            int,
+            **kwargs,
+        )
+
+        if record_creation_func in OUT_RECORD_FUNCTIONS:
+            # Ensure VAL is clamped to valid range of values.
+            # The DRVH field is a signed LONG value, but PandA uses unsigned 32-bit
+            # which can overflow it.
```

Comment:
At what point do we do the conversion to int? And if the record itself is now an analog, rather than long, doesn't that not really reflect what the underlying data in the PandA actually is?

Reply:
If we use an `ao` with `PREC=0` then the value of the record will be a float64, but displayed as rounded to the nearest integer. This means that on the GUI it will look like the value came from a longin. A float64 can display up to 56-bit integers without rounding error, so we can convert to int in `on_update` before sending to PandA. This means we won't overflow and can ditch the clamping of max values code here

Reply:
Implemented in 3e883740c4e375a167da8ea0b27ea45375c3df9c. I left the bounds checking in place, although if I understand correctly its now sort of unnecessary as the float64 is large enough to hold any value the PandA might send.